pkgs/dart_model/lib/src/json_buffer.dart

```diff
@@ -0,0 +1,390 @@
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:typed_data';
+
+/// Map with lazily evaluated values.
+///
+/// This is the efficient way to accumulate data into a [JsonBuffer]. Nested
+/// maps should also be `LazyMap`.
+class LazyMap with MapMixin<String, Object?> implements Map<String, Object?> {
+  @override
+  final List<String> keys;
+  final Object? Function(String) lookup;
+
+  LazyMap(Iterable<String> keys, this.lookup)
+      : keys = UnmodifiableListView(keys.toList());
+
+  @override
+  Object? operator [](Object? key) => key is String ? lookup(key) : null;
+
+  @override
+  void operator []=(String key, Object? value) =>
+      throw UnsupportedError('LazyMap is immutable.');
+
+  @override
+  void clear() => throw UnsupportedError('LazyMap is immutable.');
+
+  @override
+  Object? remove(Object? key) =>
+      throw UnsupportedError('LazyMap is immutable.');
+}
+
+/// Bytebuffer-backed JSON data.
+///
+/// Data can be accumulated directly into the buffer which is then immediately
+/// ready to write to the wire. On the receiving side, the data can be
+/// accessed directly as a `Map<String, Object?>` without copying out of the
+/// buffer.
+///
+/// The constructor takes a [Map], but passing in a normal `Map` _does_
+/// require copying. Instead, instantiate with a [LazyMap]. The lazy map will
+/// then be evaluated into the buffer. Nested maps in the `LazyMap` must _also_
+/// use `LazyMap` otherwise it will still be necessary to copy those.
+//
+// TODO(davidmorgan): check if it's worth adding a `LazyList` for the same
+// reason.
+// TODO(davidmorgan): this is just a proof of concept to make sure `dart_model`
+// is compatible with fast JSON. Write up a design for discussion, complete
+// the implementation.
+class JsonBuffer {
+  // TODO(davidmorgan): _seenStrings and _decodedStrings are some simple
+  // optimizations; complete the design and implementation.
+  final Map<String, _Pointer> _seenStrings = {};
+  final Map<_Pointer, String> _decodedStrings = {};
+
+  /// The JSON data.
+  // TODO(davidmorgan): what's a good initial size?
+  Uint8List _buffer = Uint8List(32);
+
+  /// The next free location to write to in the buffer.
+  int _nextFree = 0;
+
+  /// Instantiates a buffer holding [map].
+  JsonBuffer(Map<String, Object?> map) {
+    _addMap(map);
+  }
+
+  /// Immutable `Map` view of the JSON data.
+  late final Map<String, Object?> asMap = JsonBufferMap._(this, 0);
+
+  /// Instantiates using previously-serialized data.
+  ///
+  /// The buffer is _not_ copied, unpredictable behavior will result if it is
+  /// mutated.
+  JsonBuffer.deserialize(this._buffer) : _nextFree = _buffer.length;
+
+  /// The JSON data.
+  Uint8List serialize() => _buffer.sublist(0, _nextFree);
```

Comment:
Note that this does a copy, is that what we want? What about `sublistView`?

Reply:
Oh, that's what I thought I was doing ... thanks, done.